File: examples/dart/lib/billing/domain/order.dart
// lib/billing/domain/order.dart — Billing domain entities

enum OrderStatus { pending, confirmed, shipped, delivered, cancelled }

class OrderItem {
  final String productId;
  final int quantity;
  final int unitPriceCents;

  OrderItem({
    required this.productId,
    required this.quantity,
    required this.unitPriceCents,
  });
}

class Order {
  final String id;
  final String userId;
  final List<OrderItem> items;
  final OrderStatus status;
  final int totalCents;
  final DateTime createdAt;
  final DateTime updatedAt;

  Order({
    required this.id,
    required this.userId,
    required this.items,
    required this.status,
    required this.totalCents,
    required this.createdAt,
    required this.updatedAt,
  });
}

abstract class OrderRepository {
  Future<Order?> findById(String id);
  Future<void> save(Order order);
  Future<List<Order>> listByUser(String userId, int limit);
}

File: examples/typescript/src/billing/domain/order.ts
// src/billing/domain/order.ts — Billing domain entities

export interface OrderStatus {
  readonly value: string;
}

function status(value: string): OrderStatus {
  return { value };
}

export const OrderStatus = {
  Pending: status("pending"),
  Confirmed: status("confirmed"),
  Shipped: status("shipped"),
  Delivered: status("delivered"),
  Cancelled: status("cancelled"),
} as const;

export interface OrderItem {
  productId: string;
  quantity: number;
  unitPriceCents: number;
}

export interface Order extends BaseEntity {
  type: "order";
  userId: string;
  items: OrderItem[];
  status: OrderStatus;
  totalCents: number;
  createdAt: Date;
  updatedAt: Date;
}

export interface BaseEntity {
  id: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface OrderRepository {
  findById(id: string): Promise<Order | null>;
  save(order: Order): Promise<void>;
  listByUser(userId: string, limit: number): Promise<Order[]>;
}
